lms/djangoapps/staticbook/decorators.py

```diff
@@ -0,0 +1,69 @@
+"""
+Decorators for staticbook
+"""
+from functools import wraps
+from urllib.parse import urlparse
+from django.conf import settings
+from django.contrib.auth import REDIRECT_FIELD_NAME
+from django.shortcuts import resolve_url
+from common.lib.xmodule.xmodule.course_module import COURSE_VISIBILITY_PUBLIC
+from opaque_keys.edx.keys import CourseKey
+from openedx.features.course_experience import COURSE_ENABLE_UNENROLLED_ACCESS_FLAG
+
+
+def user_passes_test(test_func, login_url=None, redirect_field_name=REDIRECT_FIELD_NAME):
+    """
+    Decorator for views that checks that the user passes the given test,
+    redirecting to the log-in page if necessary. The test should be a callable
+    that takes the user object and returns True if the user passes.
+    """
+    def decorator(view_func):
+        @wraps(view_func)
+        def _wrapped_view(request, course_id, *args, **kwargs):
+            if test_func(request.user, course_id):
+                return view_func(request, course_id, *args, **kwargs)
+            path = request.build_absolute_uri()
+            resolved_login_url = resolve_url(login_url or settings.LOGIN_URL)
+            # If the login url is the same scheme and net location then just
+            # use the path as the "next" url.
+            login_scheme, login_netloc = urlparse(resolved_login_url)[:2]
+            current_scheme, current_netloc = urlparse(path)[:2]
+            if ((not login_scheme or login_scheme == current_scheme) and
+                    (not login_netloc or login_netloc == current_netloc)):
+                path = request.get_full_path()
+            from django.contrib.auth.views import redirect_to_login
+            return redirect_to_login(
+                path, resolved_login_url, redirect_field_name)
+        return _wrapped_view
+    return decorator
+
+
+def allow_anonymous_or_login_required(function=None, redirect_field_name=REDIRECT_FIELD_NAME, login_url=None):
+    """
+    Decorator for views that checks that the course content is accesable to anonymous user,
+    if not it redirecting to the log-in page if necessary.
+    """
+    def public_course(user, course_id):
+        """
+        check if the course is accessable to anonymous users
+        """
+        if user.is_authenticated:
+            return True
+        course_key = CourseKey.from_string(course_id)
+        course = get_course_by_id(course_key)
+        unenrolled_access_flag = COURSE_ENABLE_UNENROLLED_ACCESS_FLAG.is_enabled(course.id)
+        allow_access = unenrolled_access_flag and course.course_visibility in [COURSE_VISIBILITY_PUBLIC]
+        if allow_access:
+            return True
+        return False
+    
+    actual_decorator = user_passes_test(
```

Comment:
can we make it simple and move decorator logic in here instead of calling method within a method ?